ChiTech/ChiMesh/UnpartitionedMesh/unpartmesh_00a_general.cc

```diff
@@ -11,55 +11,81 @@ extern ChiTimer chi_program_timer;
+      {
+        size_t vp1 = (v<(num_verts-1))? v+1 : 0;
+
+        const auto& v0 = vertices[cell->vertex_ids[v]];
+        const auto& v1 = vertices[cell->vertex_ids[vp1]];
+
+        auto E01 = v1 - v0;
+        auto n   = E01.Cross(khat).Normalized();
+
+        if (n.Dot(v0 - cell->centroid) < 0.0)
+          ++num_negative_volume_elements;
+      }//for v
+    }
+    else if (cell->type == CellType::POLYHEDRON)
+    {
       for (auto& face : cell->faces)
       {
+        if (face.vertex_ids.size()<2)
+          throw std::logic_error(std::string(__PRETTY_FUNCTION__) +
+                                 ": cell-center-to-face check encountered face "
+                                 "with less than 2 vertices on a face, making "
+                                 "normal computation impossible.");
+
+        // Compute centroid
         chi_mesh::Vector3 face_centroid;
         for (uint64_t vid : face.vertex_ids)
           face_centroid += vertices[vid];
-        face_centroid /= double(face.vertex_ids.size());
+        face_centroid /= static_cast<double>(face.vertex_ids.size());
 
-        if (face.vertex_ids.size()<2)
-          throw std::logic_error(std::string(__PRETTY_FUNCTION__) +
-                                 ": cell-center-to-face check encountered face with less than"
-                                 " 2 vertices on a face, making normal computation impossible.");
-
-        const auto& fv1 = vertices[face.vertex_ids[0]];
-        const auto& fv2 = vertices[face.vertex_ids[1]];
+        // Form tets for each face edge
+        size_t num_face_verts = face.vertex_ids.size();
+        for (size_t fv=0; fv<face.vertex_ids.size(); ++fv)
+        {
+          size_t fvp1 = (fv<(num_face_verts-1))? fv+1 : 0;
 
-        auto E0 = fv1-face_centroid;
-        auto E1 = fv2-face_centroid;
-        auto n  = E0.Cross(E1); n.Normalize();
+          const auto& fv1 = vertices[face.vertex_ids[fv]];
+          const auto& fv2 = vertices[face.vertex_ids[fvp1]];
 
-        auto CC = face_centroid - cell->centroid;
+          auto E0 = fv1-face_centroid;
+          auto E1 = fv2-face_centroid;
+          auto n  = E0.Cross(E1).Normalized();
 
-        if (n.Dot(CC) < 0.0)
-        {
-          std::vector<uint64_t> reversed_verts(face.vertex_ids.rbegin(),
-                                               face.vertex_ids.rend());
-          face.vertex_ids = reversed_verts;
-          ++check0_corrections;
+          if (n.Dot(fv1 - cell->centroid) < 0.0)
+            ++num_negative_volume_elements;
         }
       }//for face
+    }//if polyhedron
+  }//for cell in raw_cells
 
-  if (check0_corrections > 0)
+  if (num_negative_volume_elements > 0)
     chi_log.Log(LOG_ALLWARNING)
-      << "Cell-center-to-face orientation detected " << check0_corrections
-      << " faces that violated quality requirements. An attempt to fix it"
-      << " was made.";
+      << "Cell-center-to-face orientation detected " << num_negative_volume_elements
```

Comment:
A little unsure of the meaning of cell-center-to-face orientation. What is meant by this?

Reply:
Mmmm. Help me find better words for this. When a cell-centroid looks at a face then the face should be oriented away from the centroid in order to be able to form a positive volume triangle/tet. In other words, forming a vector from the cell-centroid to the face-centroid (or any of its vertices for that matter), then the dot product with the face-normal `n` should be positive.

Reply:
I would be explicit:
- I guess it necessary applies to tri/tet for now so let's say in the error message that there is a tri/tet orientation issue
- maybe add " issue" --> orientation issue detected


alsom is this specific to PWL discretization? Would there be an issue if LD was used?

Reply:
Reworded. Message now reads like this:

```
Cell quality checks detected 3 negative volume
 sub-elements (sub-triangle or sub-tetrahedron). 
This can create issues for some spatial discretizations.
```